File: 백준/Silver/1986. 체스/체스.py
dirY = [-1,-1,-1,0,1,1,1,0]
dirX = [-1,0,1,1,1,0,-1,-1]
knightDirY = [-1,-2,-2,-1,1,2,2,1]
knightDirX = [-2,-1,1,2,2,1,-1,-2]
n,m = map(int,input().split(' '))
queen = list(map(int,input().split(' ')))
knignt = list(map(int,input().split(' ')))
pawn = list(map(int,input().split(' ')))
board = [[0 for j in range(m + 1)]for i in range(n + 1)]
for i in range(pawn[0]):
    board[pawn[i*2+1]][pawn[i*2+2]] = 2
for i in range(queen[0]):
    board[queen[i*2+1]][queen[i*2+2]] = 2
for i in range(knignt[0]):
    board[knignt[i*2+1]][knignt[i*2+2]] = 2


for i in range(queen[0]):
    y = queen[i*2+1]
    x = queen[i*2+2]
    for j in range(8):
        size = 1
        while True:
            nextY = y + dirY[j] * size
            nextX = x + dirX[j] * size
            if nextY <= 0 or nextY > n or nextX <= 0 or nextX > m or board[nextY][nextX] == 2:
                break
            board[nextY][nextX] = 1
            size += 1

for i in range(knignt[0]):
    y = knignt[i*2+1]
    x = knignt[i*2+2]
    for j in range(8):
        nextY = y + knightDirY[j]
        nextX = x + knightDirX[j]
        if nextY <= 0 or nextY > n or nextX <= 0 or nextX > m or board[nextY][nextX] == 2:
            continue
        board[nextY][nextX] = 1

ans = 0
for i in range(1,n+1):
    for j in range(1,m+1):
        if board[i][j] == 0:
           ans += 1

print(ans)

File: 백준/Silver/1986. 체스/체스.cc
#include <iostream>
#include <vector>

using namespace std;

int n, m;
int board[1001][1001];
int queenDirY[8] = { 1,1,1,0,-1,-1,-1,0 };
int queenDirX[8] = { -1,0,1,1,1,0,-1,-1 };
int knightDirY[8] = { 2,2,1,-1,-2,-2,-1,1 };
int knightDirX[8] = { -1,1,2,2,1,-1,-2,-2 };
int main()
{
	ios::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);
	vector<pair<int, int>> queen;
	vector<pair<int, int>> knight;
	cin >> n >> m;
	int a;
	cin >> a;
	for (int i = 0; i < a; ++i)
	{
		int y, x;
		cin >> y >> x;
		--y; --x;
		queen.push_back({ y,x });
		board[y][x] = 1;
	}
	cin >> a;
	for (int i = 0; i < a; ++i)
	{
		int y, x;
		cin >> y >> x;
		--y; --x;
		knight.push_back({ y,x });
		board[y][x] = 1;
	}
	cin >> a;
	for (int i = 0; i < a; ++i)
	{
		int y, x;
		cin >> y >> x;
		--y; --x;
		board[y][x] = 1;
	}
	for (int i = 0; i < queen.size(); ++i)
	{
		for (int j = 0; j < 8; ++j)
		{
			int size = 1;
			while (true)
			{
				int nextY = queen[i].first + queenDirY[j] * size;
				int nextX = queen[i].second + queenDirX[j] * size;
				if (nextY < 0 || nextY >= n || nextX < 0 || nextX >= m || board[nextY][nextX] == 1)
					break;
				board[nextY][nextX] = 2;
				++size;
			}
		}
	}
	for (int i = 0; i < knight.size(); ++i)
	{
		for (int j = 0; j < 8; ++j)
		{
			int nextY = knight[i].first + knightDirY[j];
			int nextX = knight[i].second + knightDirX[j];
			if (nextY < 0 || nextY >= n || nextX < 0 || nextX >= m || board[nextY][nextX] == 1)
				continue;
			board[nextY][nextX] = 2;
		}
	}
	int result = 0;
	for (int i = 0; i < n; ++i)
	{
		for (int j = 0; j < m; ++j)
		{
			if (board[i][j] == 0)++result;
		}
	}
	cout << result;
}